### PR TITLE
Fix payload fairing thickness calculation

### PR DESCRIPTION
**Pull request by FioraAeterna**
When the middle of the payload fairing is wider than the base, fairing thickness was incorrectly calculated. This allowed players to effectively cheese the thickness by making the base arbitrarily small, allowing for arbitrarily thin fairings.

Now, the fairing thickness should be based on the maximum width of the fairing instead of the base.

A little bit of refactoring was required to make this work, since we need the scan data earlier to determine the correct payload fairing width.

**Note to maintainers:**

1. I have not heavily tested this, just confirmed that it seems to do the right thing on a few of my rockets. Please at least look it over to make sure the code passes your smell test. I also only tested payload fairings.

2. **This will make some fairings heavier and more expensive than before**. If this effect is too large, it may be worth revisiting the minimum thickness calculations. But do note, I did not actually change the thickness formula; I simply made it apply consistently.

3. I had to refactor a few things to make this work (e.g. pre-calculating the scan to avoid doing it twice). If you don't like how I did it, please feel free to change the order of things.

### CODE REVIEW

File: Source/ProceduralFairings/FairingBase.cs [hunk 1]
```diff
@@ -139,14 +139,17 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         float lastHeight = -1000;
         float lastExtraHt = -1000;
 
+        // for non-inline fairings, the "real" maximum width of the fairing, the bulge in the middle
```
Comment (siimav) [inline Source/ProceduralFairings/FairingBase.cs]: Might was well use `///` comments before members.

File: Source/ProceduralFairings/FairingBase.cs [hunk 2]
```diff
@@ -1123,14 +1130,24 @@ private void FillProfileOutline(PayloadScan scan)
...
+        {
+            if (Mode != BaseMode.Payload)
+            {
+                maxFairingSize = baseSize;
+            }
+            else if (!autoShape)
+            {
+                maxFairingSize = manualMaxSize;
+            }
+            maxFairingSize = Mathf.Max(scan.profile.Take(scan.profile.Count).ToArray()) * 2;
```
Comment (siimav) [inline Source/ProceduralFairings/FairingBase.cs]: Might as well use `scan.profile.Max()` here.

File: Source/ProceduralFairings/FairingBase.cs [hunk 3]
```diff
@@ -537,12 +540,16 @@ public void UpdateShape(bool pushAttachments)
 
             Profiler.BeginSample("PF.UpdateShape.Setup");
             SetUIFieldLimits();
+            Profiler.BeginSample("PF.RecalcShape.ScanPayload");
+            var scan = ScanPayload();
```
Comment (siimav) [inline Source/ProceduralFairings/FairingBase.cs]: Only use `var` if the type is apparent from the right side of the assignment.

File: Source/ProceduralFairings/FairingBase.cs [hunk 4]
```diff
@@ -1123,14 +1130,24 @@ private void FillProfileOutline(PayloadScan scan)
...
 
+        private void RecalcMaxSize(PayloadScan scan)
+        {
+            if (Mode != BaseMode.Payload)
+            {
+                maxFairingSize = baseSize;
+            }
+            else if (!autoShape)
+            {
+                maxFairingSize = manualMaxSize;
```
Comment (siimav) [inline Source/ProceduralFairings/FairingBase.cs:1144]: Hmm, this will always get overwritten with the calculation 2 lines below. Probably not intended?

**Review by siimav — CHANGES_REQUESTED** [review]

File: Source/ProceduralFairings/FairingBase.cs (hunk 4 above)
Comment (FioraAeterna) [reply Source/ProceduralFairings/FairingBase.cs:1144]: Right, this is a bug. Will fix.

**Comment by FioraAeterna** [comment]
updated per comments above

File: Source/ProceduralFairings/FairingBase.cs [hunk 5]
```diff
@@ -139,14 +139,17 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         float lastHeight = -1000;
         float lastExtraHt = -1000;
 
+        /// for non-inline fairings, the "real" maximum width of the fairing, the bulge in the middle
```
Comment (siimav) [inline Source/ProceduralFairings/FairingBase.cs]: Oh, sorry. I meant this:
```
/// <summary>
/// This is a comment.
/// </summary>
```
Usually VS will insert the stub automatically once you write 3 forward slashes.

**Comment by FioraAeterna** [comment]
Looks like there might be some potential issues running this patch in flight scene; note to self to track those down.

**Comment by FioraAeterna** [comment]
Okay, hopefully this should make it work without problems in flight as well

File: Source/ProceduralFairings/FairingBase.cs [hunk 6]
```diff
@@ -94,6 +94,8 @@ public enum BaseMode { Payload, Adapter, Plate, Other }
         [KSPField] public float decouplerMassMult = 1;              // Mass multiplier
         [KSPField] public float decouplerMassBase = 0;              // Flat additional mass (0.001 = 1kg)
 
+        [KSPField] public float maxFairingSize = 0;                 // the "real" maximum width of the fairing, the bulge in the middle
```
Comment (siimav) [inline Source/ProceduralFairings/FairingBase.cs]: Needs `isPersistant = true` for the field value to get saved into the craft file / sfs.

**Comment by NathanKell** [comment]
Any updates on this?

**Comment by NathanKell** [comment]
@FioraAeterna checking in again :)

**Comment by NathanKell** [comment]
@siimav good to go?

**Pull request merged**